app/src/main/java/com/infomaniak/mail/utils/SearchUtils.kt

```diff
@@ -0,0 +1,94 @@
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.infomaniak.mail.utils
+
+import android.util.Log
+import com.infomaniak.mail.data.cache.RealmDatabase
+import com.infomaniak.mail.data.cache.mailboxContent.FolderController
+import com.infomaniak.mail.data.cache.mailboxContent.MessageController
+import com.infomaniak.mail.data.cache.mailboxContent.ThreadController
+import com.infomaniak.mail.data.models.Thread
+import com.infomaniak.mail.data.models.Thread.ThreadFilter
+import com.infomaniak.mail.data.models.message.Message
+import io.realm.kotlin.ext.toRealmList
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+
+object SearchUtils {
+
+    val TAG = SearchUtils::class.simpleName
+
+    fun searchFilters(query: String?, filters: Set<ThreadFilter>): String {
+        val filtersQuery = StringBuilder("severywhere=${if (filters.contains(ThreadFilter.FOLDER)) "0" else "1"}")
+        if (!query.isNullOrBlank()) filtersQuery.append("&scontains=$query")
+
+        with(filters) {
+            if (contains(ThreadFilter.ATTACHMENTS)) filtersQuery.append("&sattachments=yes")
+            when {
+                contains(ThreadFilter.SEEN) -> filtersQuery.append("&filters=seen")
+                contains(ThreadFilter.UNSEEN) -> filtersQuery.append("&filters=unseen")
+                contains(ThreadFilter.STARRED) -> filtersQuery.append("&filters=starred")
+                else -> Unit
+            }
+        }
+        return filtersQuery.toString()
+    }
+
+    fun selectFilter(filter: ThreadFilter, selectedFilters: MutableSet<ThreadFilter>): MutableSet<ThreadFilter> {
+        val filters = when (filter) {
+            ThreadFilter.SEEN -> {
+                selectedFilters.apply {
+                    removeAll(arrayOf(ThreadFilter.UNSEEN, ThreadFilter.STARRED))
+                }
+            }
+            ThreadFilter.UNSEEN -> {
+                selectedFilters.apply {
+                    removeAll(arrayOf(ThreadFilter.SEEN, ThreadFilter.STARRED))
+                }
+            }
+            ThreadFilter.STARRED -> {
+                selectedFilters.apply {
+                    removeAll(arrayOf(ThreadFilter.SEEN, ThreadFilter.UNSEEN))
+                }
+            }
+            else -> selectedFilters
+        }
+
+        return filters.apply { add(filter) }
+    }
```

Comment:
```suggestion
    fun selectFilter(filter: ThreadFilter, selectedFilters: MutableSet<ThreadFilter>): MutableSet<ThreadFilter> {

        val filtersToRemove = when (filter) {
            ThreadFilter.SEEN -> arrayOf(ThreadFilter.UNSEEN, ThreadFilter.STARRED)
            ThreadFilter.UNSEEN -> arrayOf(ThreadFilter.SEEN, ThreadFilter.STARRED)
            ThreadFilter.STARRED -> arrayOf(ThreadFilter.SEEN, ThreadFilter.UNSEEN)
            else -> arrayOf()
        }

        return selectedFilters.apply {
            removeAll(filtersToRemove.toSet())
            add(filter)
        }
    }
```

Reply:
🤔 I admit that it is possible to simplify, I accept your modifications but I will simplify even more 😅